# 4.7 Deriving New Variables

## Contents:

01. Import libraries
02. Import datasets¶
03. Create variables and functions
04. Creating variables with iloc
05. Searching the whole dataframe with iloc
06. For-loop
07. TASK ANSWERS FOR 4.7
08. Create new column for busiest_days
09. Comment on the new column
10. Create new column for busiest_period_of_day
11. Export dataframe

## 01. Import libraries

In [14]:
# import libraries

import pandas as pd
import numpy as np
import os

## 02. Import datasets

In [16]:
# import datasets

path = r'/Users/analazarevska/Documents/CAREER FOUNDRY/Data Analytics Program/Data Immersion/A4, Python Fundamentals for Data Analysts/Achievement 4/07-2024 Instacart Basket Analysis'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

## 03. Create variables and functions

In [17]:
# create subset from df_ords_prods

df = ords_prods_merge[:1000000]

In [18]:
df.shape

(1000000, 14)

In [22]:
# create function

def price_label_func(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) & (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return np.nan

In [24]:
# apply the function

df['price_range'] = df.apply(price_label_func, axis=1)

/var/folders/cd/qrn0mbpn33xd1226hwqc_kn00000gn/T/ipykernel_99294/3508146100.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label_func, axis=1)


In [26]:
# display number of values in price_range categories

df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    673183
Low-range product    314392
High range            12425
Name: count, dtype: int64

In [28]:
# check highest price

df['prices'].max()

99999.0

In [38]:
# check the ubnormally high price

df_high =  df[df['prices']== 99999.0]
df_high.head()

order_id  user_id  order_number  orders_day_of_week  \
128911    183964      873             3                   0   
128924   1851256      873             4                   6   
284218   1915696     1893             1                   5   
284229   2763293     1893             2                   4   
284252   2564805     1893             4                   1   

        order_hour_of_day  days_since_prior_order  product_id  \
128911                 10                     7.0       33664   
128924                 12                    13.0       33664   
284218                 17                     0.0       33664   
284229                 16                    13.0       33664   
284252                 17                    30.0       33664   

        add_to_cart_order  reordered _merge           product_name  aisle_id  \
128911                 11          0   both  2 % Reduced Fat  Milk        84   
128924                  8          1   both  2 % Reduced Fat  Milk        84   
284218                 10          0   both  2 % Reduced Fat  Milk        84   
284229                  6          1   both  2 % Reduced Fat  Milk        84   
284252                  3          1   both  2 % Reduced Fat  Milk        84   

        department_id   prices price_range  
128911             16  99999.0  High range  
128924             16  99999.0  High range  
284218             16  99999.0  High range  
284229             16  99999.0  High range  
284252             16  99999.0  High range

## 04. Creating variables with iloc

In [187]:
# create condition with iloc

df.loc[df['prices'] > 15, 'price_label'] = 'High-range product'

In [189]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_label'] = 'Mid-range product' 

In [191]:
df.loc[df['prices'] <= 5, 'price_label'] = 'Low-range product'

In [193]:
df['price_label'].value_counts(dropna = False)

price_label
Mid-range product     673183
Low-range product     314392
High-range product     12425
Name: count, dtype: int64

## 05. Searching the whole dataframe with iloc

In [48]:
# search the whole dataframe

ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_label'] = 'High-range product'

In [50]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_label'] = 'Mid-range product'

In [52]:
ords_prods_merge.loc[ords_prods_merge['prices'] <=5, 'price_label'] = 'Low-range product'

In [54]:
ords_prods_merge['price_label'].value_counts(dropna = False)

price_label
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

## 06. For-loop

In [56]:
# count values in orders_days_of_week

ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [58]:
# create for-loop function

result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [110]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

In [59]:
# create new column for the for-loop result

ords_prods_merge['busiest_day'] = result

In [61]:
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [64]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     0.0         196                  1          0   both   
1                     0.0       14084                  2          0   both   
2                     0.0       12427                  3          0   both   
3                     0.0       26088                  4          0   both   
4                     0.0       26405                  5          0   both   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

         price_label     busiest_day  
0  Mid-range product  Regularly busy  
1  Mid-range product  Regularly busy  
2  Low-range product  Regularly busy  
3  Low-range product  Regularly busy  
4  Low-range product  Regularly busy

# 07. TASK ANSWERS FOR 4.7

## 08. Create new column for busiest_days

In [66]:
# calculate value count for days

day_counts = ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

In [68]:
day_counts

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [70]:
# create for-loop function

updated_result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if (value == 0 or value == 1):
    updated_result.append("Two busiest days")
  elif (value == 4 or value == 3):
    updated_result.append("Two least busy days")
  else:
    updated_result.append("Regularly busy")

In [71]:
# create new column for the for-loop result

ords_prods_merge['busiest_days'] = updated_result

## 09. Comment on the new column

In [72]:
# get value count for the new column

ords_prods_merge['busiest_days'].value_counts()

busiest_days
Regularly busy         12916111
Two busiest days       11864412
Two least busy days     7624336
Name: count, dtype: int64

In [76]:
ords_prods_merge.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   
5   2398795        1             2                   3                  7   
6   2398795        1             2                   3                  7   
7   2398795        1             2                   3                  7   
8   2398795        1             2                   3                  7   
9   2398795        1             2                   3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     0.0         196                  1          0   both   
1                     0.0       14084                  2          0   both   
2                     0.0       12427                  3          0   both   
3                     0.0       26088                  4          0   both   
4                     0.0       26405                  5          0   both   
5                    15.0         196                  1          1   both   
6                    15.0       10258                  2          0   both   
7                    15.0       12427                  3          1   both   
8                    15.0       13176                  4          0   both   
9                    15.0       26088                  5          1   both   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                     Soda        77              7     9.0   
6                               Pistachios       117             19     3.0   
7                      Original Beef Jerky        23             19     4.4   
8                   Bag of Organic Bananas        24              4    10.3   
9               Aged White Cheddar Popcorn        23             19     4.7   

         price_label     busiest_day         busiest_days  
0  Mid-range product  Regularly busy       Regularly busy  
1  Mid-range product  Regularly busy       Regularly busy  
2  Low-range product  Regularly busy       Regularly busy  
3  Low-range product  Regularly busy       Regularly busy  
4  Low-range product  Regularly busy       Regularly busy  
5  Mid-range product  Regularly busy  Two least busy days  
6  Low-range product  Regularly busy  Two least busy days  
7  Low-range product  Regularly busy  Two least busy days  
8  Mid-range product  Regularly busy  Two least busy days  
9  Low-range product  Regularly busy  Two least busy days

In [78]:
ords_prods_merge.tail()

order_id  user_id  order_number  orders_day_of_week  \
32404854   2977660   206209            13                   1   
32404855   2977660   206209            13                   1   
32404856   2977660   206209            13                   1   
32404857   2977660   206209            13                   1   
32404858   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
32404854                 12                     7.0       14197   
32404855                 12                     7.0       38730   
32404856                 12                     7.0       31477   
32404857                 12                     7.0        6567   
32404858                 12                     7.0       22920   

          add_to_cart_order  reordered _merge  \
32404854                  5          1   both   
32404855                  6          0   both   
32404856                  7          0   both   
32404857                  8          0   both   
32404858                  9          0   both   

                                   product_name  aisle_id  department_id  \
32404854                           Tomato Paste         9              9   
32404855        Brownie Crunch High Protein Bar         3             19   
32404856  High Protein Bar Chunky Peanut Butter         3             19   
32404857    Chocolate Peanut Butter Protein Bar         3             19   
32404858    Roasted & Salted Shelled Pistachios       117             19   

          prices        price_label     busiest_day      busiest_days  
32404854     5.6  Mid-range product  Regularly busy  Two busiest days  
32404855     5.9  Mid-range product  Regularly busy  Two busiest days  
32404856     4.2  Low-range product  Regularly busy  Two busiest days  
32404857     4.9  Low-range product  Regularly busy  Two busiest days  
32404858     1.7  Low-range product  Regularly busy  Two busiest days

### Answer

There were 6204182 + 5660230 = 11864412 values for the two busiest days, which corresponds to the value counts for the new column.
There were 3840534 + 3783802 = 7624336 values for the two least busy days, which corresponds to the value counts for the new column.
There were 4496490 + 4213830 + 4205791 = 12916111 values for the other three days, which corresponds to the value counts for the new column.

## 10. Create new column for busiest_period_of_day

In [80]:
# calculate value count for hours

hour_counts = ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

In [82]:
hour_counts

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [84]:
# create for-loop function

hours_result = []

for value in ords_prods_merge["order_hour_of_day"]:
  if (value >= 9 and value <= 16):
    hours_result.append("Most orders")
  elif (value >= 2 and value <= 5):
    hours_result.append("Fewest orders")
  else:
    hours_result.append("Average orders")

In [85]:
# create new column for the for-loop result

ords_prods_merge['busiest_period_of_day'] = hours_result

In [86]:
# get value count for the new column

ords_prods_merge['busiest_period_of_day'].value_counts()

busiest_period_of_day
Most orders       21118071
Average orders    11024929
Fewest orders       261859
Name: count, dtype: int64

In [90]:
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   

    days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                      0.0         196                  1          0   both   
1                      0.0       14084                  2          0   both   
2                      0.0       12427                  3          0   both   
3                      0.0       26088                  4          0   both   
4                      0.0       26405                  5          0   both   
5                     15.0         196                  1          1   both   
6                     15.0       10258                  2          0   both   
7                     15.0       12427                  3          1   both   
8                     15.0       13176                  4          0   both   
9                     15.0       26088                  5          1   both   
10                    15.0       13032                  6          0   both   
11                    21.0         196                  1          1   both   
12                    21.0       12427                  2          1   both   
13                    21.0       10258                  3          1   both   
14                    21.0       25133                  4          0   both   

                               product_name  aisle_id  department_id  prices  \
0                                      Soda        77              7     9.0   
1   Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                       Original Beef Jerky        23             19     4.4   
3                Aged White Cheddar Popcorn        23             19     4.7   
4          XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                      Soda        77              7     9.0   
6                                Pistachios       117             19     3.0   
7                       Original Beef Jerky        23             19     4.4   
8                    Bag of Organic Bananas        24              4    10.3   
9                Aged White Cheddar Popcorn        23             19     4.7   
10                    Cinnamon Toast Crunch       121             14     4.0   
11                                     Soda        77              7     9.0   
12                      Original Beef Jerky        23             19     4.4   
13                               Pistachios       117             19     3.0   
14                    Organic String Cheese        21             16     8.6   

          price_label     busiest_day         busiest_days  \
0   Mid-range product  Regularly busy       Regularly busy   
1   Mid-range product  Regularly busy       Regularly busy   
2   Low-range product  

## 11. Export dataframe

In [92]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_new.pkl'))